### DISCUSSION TITLE
How to get predictable and consistent coloring/positions for my branches.

### DISCUSSION THREAD

**Discussion by Jarshvor** [mlange-42/git-graph#78] [Q&A] [answered]
So I got some unexpected behaviour on my repo's git-graph after a merge commit.

The best way I think I can explain it is with the following screenshot:
![1707497985](https://github.com/mlange-42/git-graph/assets/9705755/4ab4b74c-c4b8-4bda-abb3-dc10fd9cbbf0)

On the rightmost terminal, you can see that after the merge ( 75d6791 `dev-mutli_lang` into `dev`), my branch positions have been swapped.

-  `dev-multi_lang` (dark blue) is now the rightmost branch, instead of `dev` (green).
- and it is coloring the whole branch from tip to root dark blue, for `dev-multi_lang`. (instead of dev being colored as the more "senior" or "persistent" branch).

I am using a 'simple_edit' branching model with slightly tweaked colors. ->
Persistence and order were not modified from the original 'simple' model.

I'd like to know why the coloring/position changed with this last merge commit to 'dev' and not for the previous ones: like for example 7020a0c merging `dev-content_views`.

But I am beggining to suspect I need to understand well the 'order' and 'persistence' parameters and tweak them so that this does not happen.

Am I headed in the right direction? Will tweaking 'order' and 'persistence' prevent this unexpected branch/column swapping and maintain the "linearity" of my `dev` branch?

**Comment by mlange-42** [comment]
Hi @Jarshvor,
thank you for reporting!
Honestly, it is too long ago that I wrote the algorithm to be able to tell you exactly what is going on here.

You could try to add something like `"^(develop|dev)$"` to the persistence list, which could help git-graph to distinguish `dev` from your `dev-...` branches. This in turn might help to make `dev` a straight line, and avoid your issue. Persistence means how long-lived branches are. So `main`, `dev` and `publish` are more persistent than feature branches or `dev-...`.

Order is just how branches should be ordered in the graph, from left to right. Tweaking this might also help to keep things in order.

But, to be honest, it could also just be a bug.

**Reply by Jarshvor** [reply]
whoops... seems like I took too long to add my reply, and we both commented at the same time.

Not a bug, just an unexpected behavior on my part whilst using the 'simple' branching model which did not account for 'dev' persistence.

Once again, thanks for implementing these features!

**Comment by Jarshvor** [comment, marked as answer]
I understood the 'order' and 'persistence' parameters now and finally managed to get it working with:
```
persistence = [
    '^(master|main)$',
    '^(dev)$'
]
order = [
    '^tags/.*$',
    '^(master|main)$',
#   '^(dev-)*$',
    '()',
    '^(dev)$'
]
```
![1707511472](https://github.com/mlange-42/git-graph/assets/9705755/658c3668-d1dc-4348-aacf-07b368f476fb)


As the other discussions post before this points out, thanks for these features. They really make a difference.

**Reply by mlange-42** [reply]
Great that it works for you now, and good to read that people value it!

**Reply by Jarshvor** [reply]
Indeed! Still need to make time to check out [git-igitt](https://github.com/mlange-42/git-igitt). But I'm extremely happy with git-graph at the moment.
(Maybe _too happy_, as this little hiccup threw me into a fit that blinded me from seeing that the repos in the two right-most columns are indeed _exactly_ the same. Just visualized differently.)
